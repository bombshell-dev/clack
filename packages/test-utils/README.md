# `@clack/test-utils`

A bunch of useful test utiltiies for use inside clack.
