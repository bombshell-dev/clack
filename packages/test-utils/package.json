{
  "name": "@clack/test-utils",
  "private": true,
  "version": "1.0.0-alpha.1",
  "type": "module",
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
  },
  "types": "./dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/bombshell-dev/clack.git",
    "directory": "packages/test-utils"
  },
  "bugs": {
    "url": "https://github.com/bombshell-dev/clack/issues"
  },
  "homepage": "https://github.com/bombshell-dev/clack/tree/main/packages/test-utils#readme",
  "files": ["dist", "CHANGELOG.md"],
  "author": {
    "name": "James Garbutt",
    "url": "https://github.com/43081j"
  },
  "license": "MIT",
  "packageManager": "pnpm@9.14.2",
  "scripts": {
    "build": "tsc",
    "prepack": "pnpm build"
  },
  "dependencies": {
  },
  "devDependencies": {
  }
}
