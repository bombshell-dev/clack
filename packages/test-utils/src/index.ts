export { MockReadable } from './mock-readable.js';
export { MockWritable } from './mock-writable.js';

export function nextTick(): Promise<void> {
	return new Promise<void>((resolve) => {
		setTimeout(resolve, 0);
	});
}
