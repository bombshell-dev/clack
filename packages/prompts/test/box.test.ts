import { afterAll, afterEach, beforeAll, beforeEach, describe, expect, test, vi } from 'vitest';
import * as prompts from '../src/index.js';
import { MockReadable, MockWritable } from './test-utils.js';

describe.each(['true', 'false'])('box (isCI = %s)', (isCI) => {
	let originalCI: string | undefined;
	let output: MockWritable;
	let input: MockReadable;

	beforeAll(() => {
		originalCI = process.env.CI;
		process.env.CI = isCI;
	});

	afterAll(() => {
		process.env.CI = originalCI;
	});

	beforeEach(() => {
		output = new MockWritable();
		input = new MockReadable();
	});

	afterEach(() => {
		vi.restoreAllMocks();
	});

	test('renders message', () => {
		prompts.box('message', undefined, {
			input,
			output,
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders message with title', () => {
		prompts.box('message', 'some title', {
			input,
			output,
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders as wide as longest line with width: auto', () => {
		prompts.box('short\nsomewhat questionably long line', 'title', {
			input,
			output,
			width: 'auto',
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders as specified width', () => {
		prompts.box('short\nsomewhat questionably long line', 'title', {
			input,
			output,
			width: 0.5,
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('wraps content to fit within specified width', () => {
		prompts.box('foo bar'.repeat(20), 'title', {
			input,
			output,
			width: 0.5,
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders specified titlePadding', () => {
		prompts.box('message', 'title', {
			input,
			output,
			titlePadding: 6,
			width: 'auto',
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders specified contentPadding', () => {
		prompts.box('message', 'title', {
			input,
			output,
			contentPadding: 6,
			width: 'auto',
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders with prefix when includePrefix is true', () => {
		prompts.box('message', 'title', {
			input,
			output,
			includePrefix: true,
			width: 'auto',
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders truncated long titles', () => {
		prompts.box('message', 'foo'.repeat(20), {
			input,
			output,
			width: 0.2,
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('cannot have width larger than 100%', () => {
		prompts.box('message', 'title', {
			input,
			output,
			width: 1.1,
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders rounded corners when rounded is true', () => {
		prompts.box('message', 'title', {
			input,
			output,
			rounded: true,
			width: 'auto',
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders auto width with content longer than title', () => {
		prompts.box('message'.repeat(4), 'title', {
			input,
			output,
			width: 'auto',
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders auto width with title longer than content', () => {
		prompts.box('message', 'title'.repeat(4), {
			input,
			output,
			width: 'auto',
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders left aligned title', () => {
		prompts.box('message', 'title', {
			input,
			output,
			titleAlign: 'left',
			width: 'auto',
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders right aligned title', () => {
		prompts.box('message', 'title', {
			input,
			output,
			titleAlign: 'right',
			width: 'auto',
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders center aligned title', () => {
		prompts.box('message', 'title', {
			input,
			output,
			titleAlign: 'center',
			width: 'auto',
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders left aligned content', () => {
		prompts.box('message', 'title', {
			input,
			output,
			contentAlign: 'left',
			width: 'auto',
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders right aligned content', () => {
		prompts.box('message', 'title', {
			input,
			output,
			contentAlign: 'right',
			width: 'auto',
		});

		expect(output.buffer).toMatchSnapshot();
	});

	test('renders center aligned content', () => {
		prompts.box('message', 'title', {
			input,
			output,
			contentAlign: 'center',
			width: 'auto',
		});

		expect(output.buffer).toMatchSnapshot();
	});
});
