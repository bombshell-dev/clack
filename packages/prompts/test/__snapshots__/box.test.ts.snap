// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`box (isCI = false) > cannot have width larger than 100% 1`] = `
[
  "┌─title────────────────────────────────────────────────────────────────────────┐
",
  "│  message                                                                     │
",
  "└──────────────────────────────────────────────────────────────────────────────┘
",
]
`;

exports[`box (isCI = false) > renders as specified width 1`] = `
[
  "┌─title────────────────────────────────┐
",
  "│  short                               │
",
  "│  somewhat questionably long line     │
",
  "└──────────────────────────────────────┘
",
]
`;

exports[`box (isCI = false) > renders as wide as longest line with width: auto 1`] = `
[
  "┌─title──────────────────────────────┐
",
  "│  short                             │
",
  "│  somewhat questionably long line   │
",
  "└────────────────────────────────────┘
",
]
`;

exports[`box (isCI = false) > renders auto width with content longer than title 1`] = `
[
  "┌─title──────────────────────────┐
",
  "│  messagemessagemessagemessage  │
",
  "└────────────────────────────────┘
",
]
`;

exports[`box (isCI = false) > renders auto width with title longer than content 1`] = `
[
  "┌─titletitletitletitle─┐
",
  "│  message             │
",
  "└──────────────────────┘
",
]
`;

exports[`box (isCI = false) > renders center aligned content 1`] = `
[
  "┌─title──────┐
",
  "│  message   │
",
  "└────────────┘
",
]
`;

exports[`box (isCI = false) > renders center aligned title 1`] = `
[
  "┌───title────┐
",
  "│  message   │
",
  "└────────────┘
",
]
`;

exports[`box (isCI = false) > renders left aligned content 1`] = `
[
  "┌─title──────┐
",
  "│  message   │
",
  "└────────────┘
",
]
`;

exports[`box (isCI = false) > renders left aligned title 1`] = `
[
  "┌─title──────┐
",
  "│  message   │
",
  "└────────────┘
",
]
`;

exports[`box (isCI = false) > renders message 1`] = `
[
  "┌──────────────────────────────────────────────────────────────────────────────┐
",
  "│  message                                                                     │
",
  "└──────────────────────────────────────────────────────────────────────────────┘
",
]
`;

exports[`box (isCI = false) > renders message with title 1`] = `
[
  "┌─some title───────────────────────────────────────────────────────────────────┐
",
  "│  message                                                                     │
",
  "└──────────────────────────────────────────────────────────────────────────────┘
",
]
`;

exports[`box (isCI = false) > renders right aligned content 1`] = `
[
  "┌─title──────┐
",
  "│   message  │
",
  "└────────────┘
",
]
`;

exports[`box (isCI = false) > renders right aligned title 1`] = `
[
  "┌──────title─┐
",
  "│  message   │
",
  "└────────────┘
",
]
`;

exports[`box (isCI = false) > renders rounded corners when rounded is true 1`] = `
[
  "╭─title──────╮
",
  "│  message   │
",
  "╰────────────╯
",
]
`;

exports[`box (isCI = false) > renders specified contentPadding 1`] = `
[
  "┌─title──────────────┐
",
  "│      message       │
",
  "└────────────────────┘
",
]
`;

exports[`box (isCI = false) > renders specified titlePadding 1`] = `
[
  "┌──────title───────┐
",
  "│  message         │
",
  "└──────────────────┘
",
]
`;

exports[`box (isCI = false) > renders truncated long titles 1`] = `
[
  "┌─foofoofoo...─┐
",
  "│  message     │
",
  "└──────────────┘
",
]
`;

exports[`box (isCI = false) > renders with formatBorder formatting 1`] = `
[
  "[31m┌[39m[31m─[39mtitle[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m┐[39m
",
  "[31m│[39m  message   [31m│[39m
",
  "[31m└[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m┘[39m
",
]
`;

exports[`box (isCI = false) > renders with prefix when includePrefix is true 1`] = `
[
  "│ ┌─title──────┐
",
  "│ │  message   │
",
  "│ └────────────┘
",
]
`;

exports[`box (isCI = false) > wraps content to fit within specified width 1`] = `
[
  "┌─title────────────────────────────────┐
",
  "│  foo barfoo barfoo barfoo barfoo     │
",
  "│  barfoo barfoo barfoo barfoo barfoo  │
",
  "│   barfoo barfoo barfoo barfoo        │
",
  "│  barfoo barfoo barfoo barfoo barfoo  │
",
  "│   barfoo bar                         │
",
  "└──────────────────────────────────────┘
",
]
`;

exports[`box (isCI = true) > cannot have width larger than 100% 1`] = `
[
  "┌─title────────────────────────────────────────────────────────────────────────┐
",
  "│  message                                                                     │
",
  "└──────────────────────────────────────────────────────────────────────────────┘
",
]
`;

exports[`box (isCI = true) > renders as specified width 1`] = `
[
  "┌─title────────────────────────────────┐
",
  "│  short                               │
",
  "│  somewhat questionably long line     │
",
  "└──────────────────────────────────────┘
",
]
`;

exports[`box (isCI = true) > renders as wide as longest line with width: auto 1`] = `
[
  "┌─title──────────────────────────────┐
",
  "│  short                             │
",
  "│  somewhat questionably long line   │
",
  "└────────────────────────────────────┘
",
]
`;

exports[`box (isCI = true) > renders auto width with content longer than title 1`] = `
[
  "┌─title──────────────────────────┐
",
  "│  messagemessagemessagemessage  │
",
  "└────────────────────────────────┘
",
]
`;

exports[`box (isCI = true) > renders auto width with title longer than content 1`] = `
[
  "┌─titletitletitletitle─┐
",
  "│  message             │
",
  "└──────────────────────┘
",
]
`;

exports[`box (isCI = true) > renders center aligned content 1`] = `
[
  "┌─title──────┐
",
  "│  message   │
",
  "└────────────┘
",
]
`;

exports[`box (isCI = true) > renders center aligned title 1`] = `
[
  "┌───title────┐
",
  "│  message   │
",
  "└────────────┘
",
]
`;

exports[`box (isCI = true) > renders left aligned content 1`] = `
[
  "┌─title──────┐
",
  "│  message   │
",
  "└────────────┘
",
]
`;

exports[`box (isCI = true) > renders left aligned title 1`] = `
[
  "┌─title──────┐
",
  "│  message   │
",
  "└────────────┘
",
]
`;

exports[`box (isCI = true) > renders message 1`] = `
[
  "┌──────────────────────────────────────────────────────────────────────────────┐
",
  "│  message                                                                     │
",
  "└──────────────────────────────────────────────────────────────────────────────┘
",
]
`;

exports[`box (isCI = true) > renders message with title 1`] = `
[
  "┌─some title───────────────────────────────────────────────────────────────────┐
",
  "│  message                                                                     │
",
  "└──────────────────────────────────────────────────────────────────────────────┘
",
]
`;

exports[`box (isCI = true) > renders right aligned content 1`] = `
[
  "┌─title──────┐
",
  "│   message  │
",
  "└────────────┘
",
]
`;

exports[`box (isCI = true) > renders right aligned title 1`] = `
[
  "┌──────title─┐
",
  "│  message   │
",
  "└────────────┘
",
]
`;

exports[`box (isCI = true) > renders rounded corners when rounded is true 1`] = `
[
  "╭─title──────╮
",
  "│  message   │
",
  "╰────────────╯
",
]
`;

exports[`box (isCI = true) > renders specified contentPadding 1`] = `
[
  "┌─title──────────────┐
",
  "│      message       │
",
  "└────────────────────┘
",
]
`;

exports[`box (isCI = true) > renders specified titlePadding 1`] = `
[
  "┌──────title───────┐
",
  "│  message         │
",
  "└──────────────────┘
",
]
`;

exports[`box (isCI = true) > renders truncated long titles 1`] = `
[
  "┌─foofoofoo...─┐
",
  "│  message     │
",
  "└──────────────┘
",
]
`;

exports[`box (isCI = true) > renders with formatBorder formatting 1`] = `
[
  "[31m┌[39m[31m─[39mtitle[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m┐[39m
",
  "[31m│[39m  message   [31m│[39m
",
  "[31m└[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m─[39m[31m┘[39m
",
]
`;

exports[`box (isCI = true) > renders with prefix when includePrefix is true 1`] = `
[
  "│ ┌─title──────┐
",
  "│ │  message   │
",
  "│ └────────────┘
",
]
`;

exports[`box (isCI = true) > wraps content to fit within specified width 1`] = `
[
  "┌─title────────────────────────────────┐
",
  "│  foo barfoo barfoo barfoo barfoo     │
",
  "│  barfoo barfoo barfoo barfoo barfoo  │
",
  "│   barfoo barfoo barfoo barfoo        │
",
  "│  barfoo barfoo barfoo barfoo barfoo  │
",
  "│   barfoo bar                         │
",
  "└──────────────────────────────────────┘
",
]
`;
